Poprawka 5.3 (#46)
migoox/aisd-notatki: content/AlgorytmyWykladniczeZadania.tex

\paragraph{Rozwiązanie:} 

> Niech graf $G$ to graf wejściowy w naszym algorytmie taki, że

Reviewer: ```suggestion
Niech graf $G$ będzie grafem wejściowym takim, że

```

$V(G) = \{v_1, v_2, \ldots, v_n\}$. 
Utwórzmy zbiór wierzchołków

> \[V' = \{v_{ik} : v_i\in V(G) \land k \in L_{v_i}\},\]

Reviewer: ```suggestion
\[V' = \{v_{ik} : v_i\in V(G) \land k \in L_{v_i}\}\]

```

oraz zbiór krawędzi 
\[E' = \underbrace{\{v_{ik}v_{il} : v_i \in V(G) \land k, l \in L_{v_i}\}}_\text{krawędzie list} \cup 
\underbrace{\{v_{ik}v_{jk} : v_iv_j \in E(G) \land k\in L_{v_i} \cap L_{v_j}\}}_\text{krawędzie kolorów}.\]

Niech graf $G' = (V', E')$. 

Oznacza to, że każdy wierzchołek z $V(G)$ rozbijamy na dwa wierzchołki w $V'$ odpowiadające kolorom z listy (lub jeden, jeśli na liście znajduje się tylko jeden kolor) i tworzymy pomiędzy 
nimi krawędź. Pozostałe krawędzie pomiędzy dwoma wierzchołkami

> $u$ i $v$ z $G'$ tworzymy wtedy, kiedy wierzchołki 

Reviewer: ```suggestion
$u$ i $v$ z $G'$ tworzymy wtedy, kiedy wierzchołki,

```

> z których $u$ i $v$ zostały rozbite sąsiadują ze sobą

Reviewer: ```suggestion
z których $u$ i $v$ zostały rozbite, sąsiadują ze sobą

```

w grafie $G$ oraz $u$ i $v$ odpowiadają temu samemu kolorowi. 

Wystarczy rozstrzygnąć dwudzielność 

> grafu $G'$, przy pomocy algorytmu, który zastosowaliśmy w \hyperref[exc:bipart]{zadaniu 5 z tematu ,,Przeszukiwanie grafów''}. Jeśli graf $G'$ jest dwudzielny i każdy wierzchołek z $G$ ma swojego odpowiednika w \emph{tej samej} klasie dwudzielności $X$, to możemy pokolorować go listowo.

Reviewer: ```suggestion
grafu $G'$ przy pomocy algorytmu opisanego w \hyperref[exc:bipart]{zadaniu 5 z tematu ,,Przeszukiwanie grafów''}. Graf $G'$ jest dwudzielny i każdy wierzchołek z $G$ ma swojego odpowiednika w \emph{tej samej} klasie dwudzielności $X$ wtedy i tylko wtedy, gdy możemy pokolorować go listowo.

```

1. @kubazuch może link powinien być niebieski? :3
2. zmieniłem implikację na równoważność, bo to pokazujemy w kolejnych dwóch akapitach

Author: 1. nie należy to do mojego issue
2. okej

> Dzieje się tak dlatego, że dla każdego wierzchołka $v_i\in V(G)$ klasa $X$ zawiera wierzchołek $v_{ik}$ reprezentujący pokolorowanie $v_i$ na kolor $k$. Zawiera ona dokładnie jeden wierzchołek odpowiadający $v_i$ ze względu na obecność krawędzi łączących go z innymi kolorami. Jednocześnie dla każdego koloru $k$ w klasie $X$ nie sąsiadują żadne wierzchołki $v_{ik}$, $v_{jk}$, takie że $v_iv_j\in E(G)$, co wynika z dwudzielności. Kolorowanie to jest więc poprawne.

Reviewer: ```suggestion
Dzieje się tak dlatego, że dla każdego wierzchołka $v_i\in V(G)$ klasa $X$ zawiera wierzchołek $v_{ik}$ reprezentujący pokolorowanie $v_i$ na kolor $k$. Zawiera ona dokładnie jeden wierzchołek $v_{ik}$ odpowiadający $v_i$ ze względu na obecność krawędzi łączących $v_{ik}$ z $v_{il}$ dla $l \in L_{v_i}, l \neq k$ (tzn. z innymi kolorami z $L_{v_i}$). Jednocześnie z dwudzielności wynika, że dla każdego koloru $k$ w klasie $X$ nie sąsiadują żadne wierzchołki $v_{ik}$, $v_{jk}$ takie, że $v_iv_j\in E(G)$. Kolorowanie to jest więc poprawne.

```

\paragraph{Rozwiązanie:} 

Niech graf $G$ będzie grafem wejściowym takim, że

$V(G) = \{v_1, v_2, \ldots, v_n\}$. 
Utwórzmy zbiór wierzchołków
\[V' = \{v_{ik} : v_i\in V(G) \land k \in L_{v_i}\}\]

oraz zbiór krawędzi 
\[E' = \underbrace{\{v_{ik}v_{il} : v_i \in V(G) \land k, l \in L_{v_i}\}}_\text{krawędzie list} \cup 
\underbrace{\{v_{ik}v_{jk} : v_iv_j \in E(G) \land k\in L_{v_i} \cap L_{v_j}\}}_\text{krawędzie kolorów}.\]

Niech graf $G' = (V', E')$. 

Oznacza to, że każdy wierzchołek z $V(G)$ rozbijamy na dwa wierzchołki w $V'$ odpowiadające kolorom z listy (lub jeden, jeśli na liście znajduje się tylko jeden kolor) i tworzymy pomiędzy 
nimi krawędź. Pozostałe krawędzie pomiędzy dwoma wierzchołkami
$u$ i $v$ z $G'$ tworzymy wtedy, kiedy wierzchołki,

z których $u$ i $v$ zostały rozbite, sąsiadują ze sobą

w grafie $G$ oraz $u$ i $v$ odpowiadają temu samemu kolorowi. 

Wystarczy rozstrzygnąć dwudzielność 

> grafu $G'$ przy pomocy algorytmu opisanego w \hyperref[exc:bipart]{zadaniu 5 z tematu ,,Przeszukiwanie grafów''}. Graf $G'$ jest dwudzielny i każdy wierzchołek z $G$ ma swojego odpowiednika w \emph{tej samej} klasie dwudzielności $X$ wtedy i tylko wtedy, gdy możemy pokolorować go listowo.

Author: teraz to nie ma sensu w takim kontekście, że $X$ jest znikąd nazwany

\begin{figure}[H]
\centering
\includegraphics[trim=0 150 0 60,clip,width=\linewidth]{data/3listcol.eps}
\caption{Przykładowy przebieg konstrukcji nowego grafu. Można zauważyć, że powstały graf jest dwudzielny, a $v_{14}$, $v_{22}$, $v_{33}$ i $v_{41}$ tworzą żądaną klasę dwudzielności $X$, zatem oryginalny graf da się pokolorować listowo.}
\label:{fig:kolorowanielistowe}
\end{figure}

Dzieje się tak dlatego, że dla każdego wierzchołka $v_i\in V(G)$ klasa $X$ zawiera wierzchołek $v_{ik}$ reprezentujący pokolorowanie $v_i$ na kolor $k$. Zawiera ona dokładnie jeden wierzchołek $v_{ik}$ odpowiadający $v_i$ ze względu na obecność krawędzi łączących $v_{ik}$ z $v_{il}$ dla $l \in L_{v_i}, l \neq k$ (tzn. z innymi kolorami z $L_{v_i}$). Jednocześnie z dwudzielności wynika, że dla każdego koloru $k$ w klasie $X$ nie sąsiadują żadne wierzchołki $v_{ik}$, $v_{jk}$ takie, że $v_iv_j\in E(G)$. Kolorowanie to jest więc poprawne.

Pokazanie, że jeżeli istnieje poprawne kolorowanie listowe grafu $G$, to $G'$ jest dwudzielny i odpowiednie wierzchołki $G'$ odpowiadające temu kolorowaniu tworzą klasę dwudzielności jest analogiczne. 

> Utworzenie zbiorów jak i algorytm rozstrzygania dwudzielności wymagają czasu wielomianowego, więc rozwiązanie spełnia wymagania czasowe.

Reviewer 2: ```suggestion
Utworzenie zbiorów, dwudzielność oraz znajdowanie składowych grafu są problemami wymagającymi czasu wielomianowego, a zatem algorytm spełnia wymagania czasowe.
```
